aospy/calc.py

```diff
@@ -432,68 +485,71 @@ def _create_input_data_obj(self, var, start_date=False,
+                self.pfull_coord = ds[PFULL_STR]
             except KeyError:
                 pass
         return arr
 
+    def _get_pressure_from_p_coords(self, ps, name='p', n=0):
+        """Get pressure or pressure thickness array for data on p-coords."""
+        if np.any(self.pressure):
+            pressure = self.pressure
+        else:
+            pressure = self.model[n].level
+        if name == 'p':
+            return pressure
+        if name == 'dp':
+            return dp_from_p(pressure, ps)
+        raise ValueError("name must be 'p' or 'dp':"
+                         "'{}'".format(name))
+
+    def _get_pressure_from_eta_coords(self, ps, name='p', n=0):
+        """Get pressure (p) or p thickness array for data on model coords."""
+        bk = self.model[n].bk
+        pk = self.model[n].pk
+        pfull_coord = self.model[n].pfull
+        if name == 'p':
+            return pfull_from_ps(bk, pk, ps, pfull_coord)
+        if name == 'dp':
+            return dp_from_ps(bk, pk, ps, pfull_coord)
+        raise ValueError("name must be 'p' or 'dp':"
+                         "'{}'".format(name))
+
     def _get_pressure_vals(self, var, start_date, end_date, n=0):
         """Get pressure array, whether sigma or standard levels."""
-        ps = self._create_input_data_obj(self.ps, start_date, end_date)
-
+        try:
+            ps = self._ps_data
+        except AttributeError:
+            self._ps_data = self._create_input_data_obj(self.ps, start_date,
+                                                        end_date)
+            ps = self._ps_data
         if self.dtype_in_vert == 'pressure':
-            if np.any(self.pressure):
-                pressure = self.pressure
-            else:
-                pressure = self.model[n].level
-            if var.name == 'p':
-                data = pressure
-            elif var.name == 'dp':
-                data = dp_from_p(pressure, ps)
-
-        elif self.dtype_in_vert == 'sigma':
-            bk = self.model[n].bk
-            pk = self.model[n].pk
-            pfull_coord = self.model[n].pfull
-            if var.name == 'p':
-                data = pfull_from_ps(bk, pk, ps, pfull_coord)
-            elif var.name == 'dp':
-                data = dp_from_ps(bk, pk, ps, pfull_coord)
-            else:
-                raise ValueError("var.name must be 'p' or 'dp':"
-                                 "'{}'".format(var.name))
-        else:
-            raise ValueError("`dtype_in_vert` must be either 'pressure' or "
-                             "'sigma' for pressure data")
-        return data
+            return self._get_pressure_from_p_coords(ps, name=var.name, n=n)
+        if self.dtype_in_vert == ETA_STR:
+            return self._get_pressure_from_eta_coords(ps, name=var.name, n=n)
+        raise ValueError("`dtype_in_vert` must be either 'pressure' or "
+                         "'sigma' for pressure data")
 
     def _correct_gfdl_inst_time(self, arr):
```

Comment:
Good call.  Ultimately we need to think about how to package together all of the site-specific stuff (e.g. for GFDL, this; dmget/hsmget).

We could potentially explicitly check the time indices compared to the nominal range given by the input file(s), but this could be tricky to implement.